operator/pkg/reconciliation/reconcile_configsecret.go

```diff
@@ -0,0 +1,159 @@
+
+	if len(rc.Datacenter.Spec.ConfigSecret) == 0 {
+		return result.Continue()
+	}
+
+	key := types.NamespacedName{Namespace: rc.Datacenter.Namespace, Name: rc.Datacenter.Spec.ConfigSecret}
+	secret, err := rc.retrieveSecret(key)
+
+	if err != nil {
+		rc.ReqLogger.Error(err, "failed to get config secret", "ConfigSecret", key.Name)
+		return result.Error(err)
+	}
+
+	if err := rc.checkDatacenterNameAnnotation(secret); err != nil {
+		rc.ReqLogger.Error(err, "annotation check for config secret failed", "ConfigSecret", secret.Name)
+	}
+
+	config, err := getConfigFromConfigSecret(rc.Datacenter, secret)
+	if err != nil {
+		rc.ReqLogger.Error(err, "failed to get json config from secret", "ConfigSecret", rc.Datacenter.Spec.ConfigSecret)
+		return result.Error(err)
+	}
+
+	secretName := getDatacenterConfigSecretName(rc.Datacenter)
+	dcConfigSecret, exists, err := rc.getDatacenterConfigSecret(secretName)
+	if err != nil {
+		rc.ReqLogger.Error(err, "failed to get datacenter config secret")
+		return result.Error(err)
+	}
+
+	storedConfig, found := dcConfigSecret.Data["config"]
+	if !(found && bytes.Equal(storedConfig, config)) {
+		if err := rc.updateConfigHashAnnotation(dcConfigSecret); err != nil {
+			rc.ReqLogger.Error(err, "failed to update config hash annotation")
+			return result.Error(err)
+		}
+
+		rc.ReqLogger.Info("updating datacenter config secret", "ConfigSecret", dcConfigSecret.Name)
+		dcConfigSecret.Data["config"] = config
+
+		if exists {
+			if err := rc.Client.Update(rc.Ctx, dcConfigSecret); err != nil {
+				rc.ReqLogger.Error(err,"failed to update datacenter config secret", "ConfigSecret", dcConfigSecret.Name)
+				return result.Error(err)
+			}
+		}
+		if err := rc.Client.Create(rc.Ctx, dcConfigSecret); err != nil {
+			rc.ReqLogger.Error(err, "failed to create datacenter config secret", "ConfigSecret", dcConfigSecret.Name)
+			return result.Error(err)
+		}
+	}
+
+	return result.Continue()
+}
+
+// checkDatacenterNameAnnotation Checks to see if the secret has the datacenter annotation.
+// If the secret does not have the annotation, it is added, and the secret is patched. The
+// secret should be the one specifiied by ConfigSecret.
+func (rc *ReconciliationContext) checkDatacenterNameAnnotation(secret *corev1.Secret) error {
+	if v, ok := secret.Annotations[api.DatacenterAnnotation]; ok &&  v == rc.Datacenter.Name {
+		return nil
+	}
+
+	patch := client.MergeFrom(secret.DeepCopy())
+	secret.Annotations[api.DatacenterAnnotation] = rc.Datacenter.Name
+	return rc.Client.Patch(rc.Ctx, secret, patch)
+}
+
+// updateConfigHashAnnotation Adds the config hash annotation to the datacenter. The value
+// of the annotation is a hash of the secret. The datacenter is then patched. Note that the
+// secret should be the datacenter secret created by the operator, NOT the secret specified
+// by ConfigSecret.
+func (rc *ReconciliationContext) updateConfigHashAnnotation(secret *corev1.Secret) error {
```

Comment:
How do we ensure that the copy secret has the same security as the original secret? If the data is sensitive, we don't want this copy to be viewable anymore than the real secret. Or if the secret came from Hashicorp Vault or some other injection method (I don't know how the customer handles them if they're really sensitive).

Reply:
>How do we ensure that the copy secret has the same security as the original secret? If the data is sensitive, we don't want this copy to be viewable anymore than the real secret. 

Great question. I think this falls more on the administrator or whoever is setting up RBAC for users and service accounts. The operator already has permissions to read/write secrets in the namespace in which it is deployed. I think handling other RBAC falls outside the scope of the operator.